src/Features/CSharp/Portable/ExtractMethod/CSharpMethodExtractor.CSharpCodeGenerator.cs

```diff
@@ -716,6 +748,65 @@ static bool ReturnOperationBelongsToMethod(SyntaxNode returnOperationSyntax, Syn
 
                     return enclosingMethod == methodSyntax;
                 }
+
+                async Task<SemanticDocument> CheckReturnOperations(SyntaxNode node)
+                {
+                    var methodOperation = semanticModel.GetOperation(node, cancellationToken);
+                    var returnOperations = methodOperation.DescendantsAndSelf().OfType<IReturnOperation>();
+
+                    foreach (var returnOperation in returnOperations)
+                    {
+                        // If the return statement is located in a nested local function or lambda it
+                        // shouldn't contribute to the nullability of the extracted method's return type
+                        if (!ReturnOperationBelongsToMethod(returnOperation.Syntax, methodOperation.Syntax))
+                        {
+                            continue;
+                        }
+
+                        var syntax = returnOperation.ReturnedValue?.Syntax ?? returnOperation.Syntax;
+                        var returnTypeInfo = semanticModel.GetTypeInfo(syntax, cancellationToken);
+                        if (returnTypeInfo.Nullability.FlowState == NullableFlowState.MaybeNull)
+                        {
+                            // Flow state shows that return is correctly nullable
+                            return await base.UpdateMethodAfterGenerationAsync(originalDocument, methodSymbolResult, cancellationToken).ConfigureAwait(false);
+                        }
+                    }
+
+                    return null;
+                }
+
+                async Task<Document> GenerateNewDocument(TypeSyntax returnType)
+                {
+                    // Return type can be updated to not be null
+                    var newType = methodSymbolResult.Data.ReturnType.WithNullability(NullableAnnotation.NotAnnotated);
+
+                    var oldRoot = await originalDocument.Document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                    var newRoot = oldRoot.ReplaceNode(returnType, newType.GenerateTypeSyntax());
+
+                    return originalDocument.Document.WithSyntaxRoot(newRoot);
+                }
+            }
+
+            protected ImmutableHashSet<string> GetLocalFunctionNamesIfScopeIsMethod(SyntaxNode scope)
```

Comment:
Do we need to provide these names if we're not generating a local function?

Reply:
If extracting code from a method body, then I believe we should keep track of local function names in the method. Otherwise the new method could be generated with the same name as an existing local function, which might be confusing to the user.

Reply:
so my overall point is: thee *existing* function we have to generate a unique, non-colliding, name should just work.  It itself may need to be updated to know about local functions.

i.e. we already have ISemanticFactsService.GenerateUniqueName.  It should already be something that can be used for something like this.  :)

Reply:
Got it, just pushed a change to fix this.